spring/boot-actuator-autoconfigure/src/main/java/com/linecorp/armeria/spring/actuate/WebOperationHttpService.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.spring.actuate;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.Nullable;
+
+import org.springframework.boot.actuate.endpoint.InvocationContext;
+import org.springframework.boot.actuate.endpoint.SecurityContext;
+import org.springframework.boot.actuate.endpoint.web.WebEndpointResponse;
+import org.springframework.boot.actuate.endpoint.web.WebOperation;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+/**
+ * {@link HttpService} to handle a {@link WebOperation}. Mostly inspired by reactive implementation in
+ * {@link org.springframework.boot.actuate.endpoint.web.reactive.AbstractWebFluxEndpointHandlerMapping}.
+ */
+final class WebOperationHttpService implements HttpService {
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final TypeReference<Map<String, Object>> JSON_MAP =
+            new TypeReference<Map<String, Object>>() {};
+
+    private final WebOperation operation;
+
+    WebOperationHttpService(WebOperation operation) {
+        this.operation = operation;
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) {
+        final CompletableFuture<HttpResponse> resFuture = new CompletableFuture<>();
+        req.aggregate().thenAccept(msg -> {
+            if (operation.isBlocking()) {
+                ctx.blockingTaskExecutor().execute(() -> invoke(ctx, msg, resFuture));
+            } else {
+                invoke(ctx, msg, resFuture);
+            }
+        });
+        return HttpResponse.from(resFuture);
+    }
+
+    private void invoke(ServiceRequestContext ctx,
+                        AggregatedHttpMessage msg,
+                        CompletableFuture<HttpResponse> resFuture) {
+        final Map<String, Object> arguments = getArguments(ctx, msg);
+        final Object result = operation.invoke(new InvocationContext(SecurityContext.NONE, arguments));
```

Comment:
Not sure if this is a TODO or not. I guess actuator doesn't do basic auth anymore due to a different approach (controlling which endpoints are exposed at all) https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-monitoring.html cc @shakuzen who is smarter than me

Reply:
From what I could understand, actuator is meant to integrate with spring security for authentication. I don't know if it's possible to build spring security on top of armeria but think it's probably overkill - if a user needs to go that far, they can go ahead and use armeria's spring webflux integration to get all spring features for free.

Would it be a blocker for zipkin-server if this actuator integration isn't connected to spring security?

Reply:
Also filed #1579 which would make it easier for users to customize the authentication in an armeria-based way.